Getrange v1

nemanja-boric-sociomantic opened the pull request.

nemanja-boric-sociomantic commented:
Blocked on making it work on DLS node.

gavin-norman-sociomantic commented on src/dlsproto/client/internal/SharedResources.d:
```diff
@@ -46,6 +49,22 @@ public final class SharedResources
+
+    ***************************************************************************/
+
+    private FreeList!(RequestEventDispatcher) request_event_dispatchers;
```
For the future: It occurs to me that we might be able to adapt `RequestEventDispatcher` to accept `void[]`s to use internally, doing away with the need to have a separate pool of them. (I think the only thing it contains that should be reused is a couple of arrays.)

nemanja-boric-sociomantic replied:
Yes. I will open the swarm issue about this.

nemanja-boric-sociomantic replied:
https://github.com/sociomantic-tsunami/swarm/issues/186

gavin-norman-sociomantic commented on src/dlsproto/client/internal/SharedResources.d:
```diff
@@ -62,6 +81,29 @@ public final class SharedResources
+    /***************************************************************************
+
+        A SharedResource instance is stored in the ConnectionSet as an
+        Object. This helper function safely casts from thsi Object to a
```
Typo thsi

gavin-norman-sociomantic commented on src/dlsproto/client/mixins/NeoSupport.d:
```diff
@@ -19,7 +19,7 @@ import ocean.core.Tuple;
     Template to strip the notifier from the variadic arguments.
 
     Params:
-        TypeToErase = type to erase from options 
```
Could you break out the whitespace changes into a separate commit?

gavin-norman-sociomantic commented on src/dlsproto/client/request/internal/GetRange.d:
```diff
@@ -62,161 +56,43 @@ public struct GetRange
+    import ocean.transition;
+    mixin TypeofThis!();
+
+    mixin BatchController!(typeof(this), IController);
```
Comment maybe? It's easy to miss this among the imports, otherwise.

gavin-norman-sociomantic commented on src/dlsproto/client/request/internal/GetRange.d:
```diff
@@ -317,422 +160,631 @@ public struct GetRange
     ***************************************************************************/
 
-    private State stateInitialising ( )
+    private class RecordStream
```
I guess these could be scope classes.

gavin-norman-sociomantic replied:
By the way, I presume that everything in here is completely generic (i.e. identical to DMQ Consume) except the extraction of records from the batch.

nemanja-boric-sociomantic replied:
```
./submodules/dlsproto/src/dlsproto/client/request/internal/GetRange.d(648): Error: variable dlsproto.client.request.internal.GetRange.GetRangeHandler.Reader.record_stream globals, statics, fields, ref and out parameters cannot be scope
```

> By the way, I presume that everything in here is completely generic (i.e. identical to DMQ Consume) except the extraction of records from the batch.

Yes, as the batches contains keys, and some few things (accepting the filter string and resuming when the new data arrives).

gavin-norman-sociomantic commented on src/dlsproto/client/request/internal/GetRange.d:
```diff
@@ -317,422 +160,631 @@ public struct GetRange
+                enableStomping(*this.batch_buffer);
+
+            // Append record_batch, then set this.remaining_batch to reference
+            // the remaining records. To void a dangling slice if
```
Typo void.

gavin-norman-sociomantic commented on src/dlsproto/client/request/internal/GetRange.d:
```diff
@@ -317,422 +160,631 @@ public struct GetRange
+                    this.passRecordToUser(
+                        *this.outer.conn.message_parser.getValue!(time_t)(
+                            this.remaining_batch),
+                        this.outer.conn.message_parser.getArray!(Const!(void))(
```
Hmmm, do we want to do away with the batch compression? DMQ Consume never had it, but here we're losing something.

nemanja-boric-sociomantic replied:
No, I think we should do the batch compression, thank you for reminding me.

gavin-norman-sociomantic commented on src/dlsproto/client/request/internal/GetRange.d:
```diff
@@ -317,422 +160,631 @@ public struct GetRange
+            {
+                auto event = this.outer.request_event_dispatcher.nextEvent(
+                    this.fiber,
+                    Signal(ControllerSignal.Resume),
```
What handles `controller.suspend`?

nemanja-boric-sociomantic replied:
Nothing, we're not expecting this signal asynchronously, but we're checking it before passing each record to the user:

https://github.com/sociomantic-tsunami/dlsproto/pull/39/files#diff-c6938f346e8e72b5984b4ebffc36d024R505

gavin-norman-sociomantic commented on src/dlsproto/node/neo/request/GetRange.d:
```diff
@@ -266,291 +188,279 @@ public abstract scope class GetRangeProtocol_v0
-
-        Returns:
-            `true` if the range preparation was sucessfull
+        Request main loop. Pops records from the storage, sends them to the
```
Pops? ;)

gavin-norman-sociomantic commented on src/dlsproto/node/neo/request/GetRange.d:
```diff
@@ -266,291 +188,279 @@ public abstract scope class GetRangeProtocol_v0
-        Retrieve the next record from the channel, if available.
+                switch (event.active)
+                {
+                    // resumed from the node, resume the iteration
```
"from the node" is a bit odd. This whole module is "the node", or?

gavin-norman-sociomantic commented on src/dlsproto/node/neo/request/GetRange.d:
```diff
@@ -266,291 +188,279 @@ public abstract scope class GetRangeProtocol_v0
-
-        Throws:
-            `ProtocolError` if `msg_type` is not a control message type.
+            `true` if there was a record in the channel, false if the channel is
```
The doc of this method seems outdated. "channel is empty"?

gavin-norman-sociomantic reviewed (commented):
Not feeling that mentally acute today. Just a few comments.

nemanja-boric-sociomantic commented:
I'll address your comments, but just unblocking this as I've confirmed this works with the node (after 4f905e1 commit being applied).

nemanja-boric-sociomantic committed:
Don't resume the Task on node_error in Blocking task support

nemanja-boric-sociomantic commented:
Updated with the comments addressed and with the batch compression.

gavin-norman-sociomantic commented:
Ooh, I just remembered a trick David used recently in dmqproto that made reviewing _way_ easier when doing a complete request rewrite like this. `version(none)` the old handler (v0) and add the new handler (v1) at the top of the module. Then add a second commit that removes the versioned out code.

nemanja-boric-sociomantic commented:
Ah, good idea, I will update.

nemanja-boric-sociomantic committed:
Migrate GetRange to multi-fiber-framework

This migrates the GetRange request from the
SuspendableRequest framework to the new multi-fiber framework.

nemanja-boric-sociomantic committed:
Remove GetRange v0 protocol

nemanja-boric-sociomantic committed:
Fix whitespace

nemanja-boric-sociomantic commented:
Updated.

david-eckardt-sociomantic commented on src/dlsproto/node/neo/request/GetRange.d:
```diff
@@ -12,13 +12,527 @@
+                    // Resumed by the storage engine, now ready to return value
+                    // with `getNextRecord`, so we try again.
+                    case event.active.resumed:
+                        continue main_loop;
```
The `main_loop` label is not needed.

nemanja-boric-sociomantic replied:
Yeah, it's _continue_ not _break_ :face_with_head_bandage:

nemanja-boric-sociomantic replied:
Updated.

david-eckardt-sociomantic commented on src/dlsproto/node/neo/request/GetRange.d:
```diff
@@ -12,13 +12,527 @@
+
+                        if (msg_type == msg_type.Ack)
+                        {
+                            acked = true;
```
Suggest replacing this variable with a `while (true)` loop and `break` here.

nemanja-boric-sociomantic replied:
To be honest, I find the current approach more verbose (in a good way). I can still change it, if you insist.
